filebeat/input/filestream/prospector_test.go

```diff
@@ -156,14 +255,98 @@ func TestProspectorDeletedFile(t *testing.T) {
 	}
 }
 
+func TestProspectorRenamedFile(t *testing.T) {
+	testCases := map[string]struct {
+		events                   []loginp.FSEvent
+		trackRename              bool
+		closeRenamed             bool
+		expectedEncounteredNames []string
+		expectedStoppedNames     []string
+	}{
+		"one renamed file without rename tracker": {
+			events: []loginp.FSEvent{
+				loginp.FSEvent{
+					Op:      loginp.OpRename,
+					OldPath: "/old/path/to/file",
+					NewPath: "/new/path/to/file",
+				},
+			},
+			expectedEncounteredNames: []string{"path::/new/path/to/file"},
+			expectedStoppedNames:     []string{"path::/old/path/to/file"},
+		},
+		"one renamed file with rename tracker": {
+			events: []loginp.FSEvent{
+				loginp.FSEvent{
+					Op:      loginp.OpRename,
+					OldPath: "/old/path/to/file",
+					NewPath: "/new/path/to/file",
+				},
+			},
+			trackRename: true,
+		},
+		"one renamed file with rename tracker with close renamed": {
+			events: []loginp.FSEvent{
+				loginp.FSEvent{
+					Op:      loginp.OpRename,
+					OldPath: "/old/path/to/file",
+					NewPath: "/new/path/to/file",
+				},
+			},
+			trackRename:          true,
+			closeRenamed:         true,
+			expectedStoppedNames: []string{"path::/old/path/to/file"},
+		},
+	}
+
+	for name, test := range testCases {
+		test := test
+
+		t.Run(name, func(t *testing.T) {
+			p := fileProspector{
+				filewatcher:       &mockFileWatcher{events: test.events},
+				identifier:        mustPathIdentifier(test.trackRename),
+				stateChangeCloser: stateChangeCloserConfig{Renamed: test.closeRenamed},
+			}
+			ctx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+
+			testStore := newMockMetadataUpdater()
+			testStore.set("path::/old/path/to/file")
+
+			hg := getTestHarvesterGroup()
+			p.Run(ctx, testStore, hg)
+
+			has := testStore.has("path::/old/path/to/file")
+			if test.trackRename {
+				assert.True(t, has)
+			} else {
+				assert.False(t, has)
+			}
+
+			assert.ElementsMatch(t, test.expectedEncounteredNames, hg.encounteredNames)
+			assert.ElementsMatch(t, test.expectedStoppedNames, hg.stoppedNames)
```

Comment:
These tests seem to aim for checking behavior and not state. The `hg` is collecting the output and the mockFileWatcher is the input. Given that input events have some order, I would expect that the order in the output is important as well. How about storing the `harvesterGroup` calls as an event stream and verifying it?

e.g.

```
type harvesterEvent interface{ String() string }
type harvesterStart string
type harvesterStop string
type harvesterGroupStop struct {}



type testHarvesterGroup struct {
  events []harvesterEvent
}

...

want := []harvesterEvent{
  harvesterStart("..."),
  harvesterStop("..."),
}
```